mlflow/pyfunc/__init__.py

```diff
@@ -370,62 +370,149 @@ def all_ints(xs):
+    the schema or if it can be safely converted to match the input schema.
+
+    For tensor-based signatures, we make sure the shape and type of the input matches the shape
+    and type specified in model's input schema.
+    """
+    if not input_schema.is_tensor_spec():
+        if isinstance(pfInput, (list, np.ndarray, dict)):
+            try:
+                pfInput = pandas.DataFrame(pfInput)
+            except Exception as e:
+                message = (
+                    "This model contains a column-based signature, which suggests a DataFrame"
+                    " input. There was an error casting the input data to a DataFrame:"
+                    " {0}".format(str(e))
+                )
+                raise MlflowException(message)
+        if not isinstance(pfInput, pandas.DataFrame):
+            message = "Expected input to be DataFrame or list. Found: %s" % type(pfInput).__name__
             raise MlflowException(message)
-    if not isinstance(pdf, pandas.DataFrame):
-        message = "Expected input to be DataFrame or list. Found: %s" % type(pdf).__name__
-        raise MlflowException(message)
 
     if input_schema.has_input_names():
         # make sure there are no missing columns
-        col_names = input_schema.input_names()
-        expected_names = set(col_names)
-        actual_names = set(pdf.columns)
-        missing_cols = expected_names - actual_names
-        extra_cols = actual_names - expected_names
+        input_names = input_schema.input_names()
+        expected_cols = set(input_names)
+        actual_cols = set()
+        if len(expected_cols) == 1 and isinstance(pfInput, np.ndarray):
+            # for schemas with a single column, match input with column
+            pfInput = {input_names[0]: pfInput}
+            actual_cols = expected_cols
+        elif isinstance(pfInput, pandas.DataFrame):
+            actual_cols = set(pfInput.columns)
+        elif isinstance(pfInput, dict):
+            actual_cols = set(pfInput.keys())
+        missing_cols = expected_cols - actual_cols
+        extra_cols = actual_cols - expected_cols
         # Preserve order from the original columns, since missing/extra columns are likely to
         # be in same order.
-        missing_cols = [c for c in col_names if c in missing_cols]
-        extra_cols = [c for c in pdf.columns if c in extra_cols]
+        missing_cols = [c for c in input_names if c in missing_cols]
+        extra_cols = [c for c in actual_cols if c in extra_cols]
         if missing_cols:
             message = (
-                "Model input is missing columns {0}."
-                " Note that there were extra columns: {1}".format(missing_cols, extra_cols)
+                "Model is missing inputs {0}."
+                " Note that there were extra inputs: {1}".format(missing_cols, extra_cols)
             )
             raise MlflowException(message)
```

Comment:
After this line, I think it would help readability and modularity to branch off based on whether the the schema defines a TensorSpec/ColSpec.
Otherwise it feels the enforcement is getting interleaved. 
Something like:
`new_pyInput = _enforce_tensor_spec(pfInput) if input_schema.is_tensor_spec() else _enforce_col_spec(pfInput).`